delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/resources/ValidateResource.scala

```diff
@@ -58,24 +59,18 @@ object ValidateResource {
           jsonld: JsonLdAssembly,
           schemaClaim: SchemaClaim,
           enforceSchema: Boolean
-      ): IO[ValidationResult] =
-        if (isUnconstrained(schemaClaim))
-          IO.raiseWhen(enforceSchema)(SchemaIsMandatory(schemaClaim.project)) >>
-            assertNotReservedId(jsonld.id) >>
-            IO.pure(NoValidation(schemaClaim.project))
-        else
-          for {
-            schema <- resolveSchema(schemaClaim)
-            result <- apply(jsonld, schema)
-          } yield result
+      ): IO[ValidationResult] = {
+        val submitOnDefinedSchema: SubmitOnDefinedSchema = resolveSchema(_, _, _).flatMap(apply(jsonld, _))
+        assertNotReservedId(jsonld.id) >> schemaClaim.validate(enforceSchema)(submitOnDefinedSchema)
```

Comment:
I think this pattern is not helping make clear what is going on here. I would suggest either `schemaClaim.validate` should return something that could be mapped on, or else just combine the functions. This callback style is really difficult to grok (despite what is happening being quite simple)

Reply:
I don't think this is equivalent to `IO.whenA`, as I don't have a problem with that. There are other patterns possible which would keep the validation logic in the same place. You don't have to do it, I'm just pointing out that this pattern makes the code harder to navigate and I don't think I'm the only one who would think this

Reply:
They are both callbacks to me :)